### PR TITLE
valset: store and expose the start of a jailing period

### LINKED ISSUE #94: `JailingPeriod` - add information about start period
Customer requirement: "The oversight community needs should be able to see a list of currently jailed validators and from when to when they are jailed"

Refactor `JailingPeriod` [implementation](https://github.com/confio/poe-contracts/blob/main/contracts/tgrade-valset/src/msg.rs#L350-L353), so it could contain information about both start and end Timestamps.

### PR DESCRIPTION
**Pull request by uint**
Closes #94

Working on migration code for this.

### CODE REVIEW

File: contracts/tgrade-valset/src/msg.rs [hunk 1]
```diff
@@ -395,25 +395,38 @@ impl OperatorResponse {
...
+    pub end: JailingEnd,
+}
+
 #[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
 #[serde(rename_all = "snake_case")]
-pub enum JailingPeriod {
+pub enum JailingEnd {
     Until(Expiration),
     Forever {},
 }
```
Comment (ueco-jb) [inline contracts/tgrade-valset/src/msg.rs:409]: :+1: I like it.

**Comment by uint** [comment]
I added migration code for this following [the pattern I found in `trusted-circle`](https://github.com/confio/tgrade-contracts/blob/main/contracts/tgrade-trusted-circle/src/migration.rs). I didn't write tests for that as that doesn't seem trivial. I assume we'll test it before an actual release.

File: contracts/tgrade-valset/src/migration.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    } else {
+        return Ok(());
+    };
+
+    for (addr, jailing_period) in jailings {
+        JAIL.save(deps.storage, &addr, &jailing_period)?;
+    }
+
+    Ok(())
+}
```
Comment (ueco-jb) [inline contracts/tgrade-valset/src/migration.rs:48]: Throw some tests here?

**Review by ueco-jb — APPROVED** [review]
lgtm

File: contracts/tgrade-valset/src/migration.rs (hunk 2 above)
Comment (uint) [reply contracts/tgrade-valset/src/migration.rs:48]: Okay, I managed to mock old state using `mock_dependencies()`, but I think that's a very incomplete smoke test.

**Comment by uint** [comment]
Rebased

File: contracts/tgrade-valset/src/contract.rs [hunk 3]
```diff
@@ -303,19 +305,17 @@ fn execute_unjail<Q: CustomQuery>(
...
-        Ok(Some(JailingPeriod::Forever {})) => {
+    // if this is `None`, the validator was not unjailed and unjailing succeeds
+    if let Some(expiration) = JAIL.may_load(deps.storage, operator)? {
+        if expiration.is_forever() {
             return Err(ContractError::UnjailFromJailForeverForbidden {});
         }
-        // Jailing period expired or called by admin - can unjail
-        Ok(Some(expiration)) if (expiration.is_expired(&env.block) || is_admin) => {
+
+        if expiration.is_expired(&env.block) || is_admin {
```
Comment (maurolacy) [inline contracts/tgrade-valset/src/contract.rs:314]: Nit: revert  these conditions for saving some processing (sometimes).

File: contracts/tgrade-valset/src/msg.rs [hunk 4]
```diff
@@ -395,25 +395,38 @@ impl OperatorResponse {
     }
 }
 
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+pub struct JailingPeriod {
+    pub start: Timestamp,
+    pub end: JailingEnd,
+}
```
Comment (maurolacy) [inline contracts/tgrade-valset/src/msg.rs:402]: I can't help but notice that these could be a pair of `u64` values, with the convention that `u64::MAX` (or `0`) means forever.

But, who am I to judge? :-)

**Pull request merged**